sn_client/src/audit/dag_crawling.rs

```diff
@@ -62,26 +64,93 @@ impl Client {
+                    let addr = spend.address();
+                    dag.insert(addr, spend.clone());
+                    if let Some(sender) = &spend_processing {
+                        sender
+                            .send(spend)
+                            .await
+                            .map_err(|e| WalletError::SpendProcessing(e.to_string()))?;
+                    }
+                }
+                Ok(dag)
+            });
+
+        // wait for both to finish
+        let (crawl_res, build_res) = tokio::join!(crawl_handle, build_handle);
+        crawl_res.map_err(|e| {
+            WalletError::SpendProcessing(format!("Failed to Join crawling results {e}"))
+        })??;
+        let mut dag = build_res.map_err(|e| {
+            WalletError::SpendProcessing(format!("Failed to Join DAG building results {e}"))
+        })??;
+
+        // verify the DAG
+        if verify {
+            info!("Now verifying SpendDAG from {spend_addr:?} and recording errors...");
+            let start = std::time::Instant::now();
+            if let Err(e) = dag.record_faults(&dag.source()) {
+                let s = format!(
+                    "Collected DAG starting at {spend_addr:?} is invalid, this is probably a bug: {e}"
+                );
+                error!("{s}");
+                return Err(WalletError::Dag(s));
+            }
+            let elapsed = start.elapsed();
+            info!("Finished verifying SpendDAG from {spend_addr:?} in {elapsed:?}");
+        }
+
+        Ok(dag)
+    }
+
+    /// Crawls the Spend Dag from a given SpendAddress recursively
+    /// following descendants all the way to UTXOs
+    /// Returns the UTXOs reached
+    pub async fn spend_dag_crawl_from(
+        &self,
+        spend_addr: SpendAddress,
+        spend_processing: Sender<SignedSpend>,
+    ) -> WalletResult<BTreeSet<SpendAddress>> {
+        info!("Crawling spend DAG from {spend_addr:?}");
+        let mut utxos = BTreeSet::new();
 
         // get first spend
         let first_spend = match self.crawl_spend(spend_addr).await {
             InternalGetNetworkSpend::Spend(s) => *s,
             InternalGetNetworkSpend::DoubleSpend(s1, s2) => {
-                dag.insert(spend_addr, *s2);
+                spend_processing
+                    .send(*s2)
+                    .await
+                    .map_err(|e| WalletError::SpendProcessing(e.to_string()))?;
                 *s1
             }
             InternalGetNetworkSpend::NotFound => {
                 // the cashnote was not spent yet, so it's an UTXO
                 info!("UTXO at {spend_addr:?}");
-                return Ok(dag);
+                utxos.insert(spend_addr);
+                return Ok(utxos);
             }
             InternalGetNetworkSpend::Error(e) => {
                 return Err(WalletError::FailedToGetSpend(e.to_string()));
             }
         };
-        dag.insert(spend_addr, first_spend.clone());
+        spend_processing
```

Comment:
not sure if we still need that `gen` stuff.

anyway, for beta_track optimisation, we don't need to track down utxos of royalty and forwarded payment
which, seems need to have a separate feature guarded function with this infrastructure.

Reply:
Agreed, `gen` is a relic from the past, we could get rid of it.

> seems need to have a separate feature guarded function with this infrastructure.

royalties/dag/beta rewards all can go through the same process receiving spends through the spend_processing channel. `sn_client` is lib code so it can stay agnostic IMO.